worlds/marioland2/logic.py

```diff
@@ -0,0 +1,608 @@
+
+
+def has_level_progression(state, item, player, count=1):
+    return state.count(item, player) + (state.count(item + " x2", player) * 2) >= count
+
+
+def mushroom_zone_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Mushroom Zone")
+    reachable_coins = 38
+    if state.has_any(["Mushroom", "Fire Flower"], player) or not auto_scroll:
+        # Was able to get all but 1, being lenient.
+        reachable_coins += 2
+    if has_pipe_down(state, player):
+        # There's 24 in each of the underground sections.
+        # The first one requires missing some question mark blocks if auto scrolling (the last +4).
+        # If you go in the second without pipe up, you can get everything except the last 5 plus the ones in the first
+        # underground section.
+        reachable_coins += 19
+        if has_pipe_up(state, player) or not auto_scroll:
+            reachable_coins += 5
+        if has_pipe_up(state, player):
+            reachable_coins += 20
+            if not auto_scroll:
+                reachable_coins += 4
+    return coins <= reachable_coins
+
+
+def tree_zone_1_normal_exit(state, player):
+    return has_pipe_right(state, player) or state.has("Tree Zone 2 Midway Bell", player),
+
+
+def tree_zone_1_coins(state, player, coins):
+    return coins <= 87 or not is_auto_scroll(state, player, "Tree Zone 1")
+
+
+def tree_zone_2_secret_exit(state, player):
+    return has_pipe_right(state, player) and state.has("Carrot", player)
+
+
+def tree_zone_2_midway_bell(state, player):
+    return has_pipe_right(state, player) or state.has("Tree Zone 2 Midway Bell", player)
+
+
+def tree_zone_2_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Tree Zone 2")
+    reachable_coins = 18
+    if has_pipe_right(state, player):
+        reachable_coins += 38
+        if state.has("Carrot", player):
+            reachable_coins += 12
+            if not auto_scroll:
+                reachable_coins += 30
+    elif state.has("Tree Zone 2 Midway Bell", player):
+        reachable_coins = 30
+        if not auto_scroll:
+            reachable_coins += 8
+    return coins <= reachable_coins
+
+
+def tree_zone_3_normal_exit(state, player):
+    return not is_auto_scroll(state, player, "Tree Zone 3")
+
+
+def tree_zone_3_coins(state, player, coins):
+    if is_auto_scroll(state, player, "Tree Zone 3"):
+        return coins <= 4
+    if coins <= 19:
+        return True
+    elif state.has_any(["Mushroom", "Fire Flower"], player) and coins <= 21:
+        return True
+    return state.has("Carrot", player)
+
+
+def tree_zone_4_normal_exit(state, player):
+    if has_pipe_down(state, player):
+        return tree_zone_4_midway_bell(state, player)
+    return False
+
+
+def tree_zone_4_midway_bell(state, player):
+    return ((has_pipe_right(state, player) and has_pipe_up(state, player))
+            or state.has("Tree Zone 4 Midway Bell", player))
```

Comment:
Could do this, which matches most other sections
```suggestion
    return has_pipe_right(state, player) and has_pipe_up(state, player) or state.has("Tree Zone 4 Midway Bell", player)
```